src/compiler/interface.jl

```diff
@@ -171,6 +186,41 @@ function copy!(x::AbstractVector,  gs::Grads)
   x
 end
 
+broadcasted(f, gss::Grads...) = map(f, gss...)
+
+for op in (:*, :/)
+  @eval broadcasted(::typeof($op), a::Number, gs::Grads) = map(x -> $op(a, x), gs)
+  @eval broadcasted(::typeof($op), gs::Grads, a::Number) = map(x -> $op(x, a), gs)
+end
+
+function materialize!(gs1::Grads, gs2::Grads)
+  issetequal(gs1.params, gs2.params) || 
+    throw(ArgumentError("Expected Grads objects with the same Params."))
+  for p in gs1.params
+    gs1[p] = gs2[p]
+  end
+  return gs1
+end
+
+function Base.map(f, gs1::Grads, gss::Grads...)
+  gsout = Grads(IdDict{Any,Any}(), Params(gs1.params))
+  return map!(f, gsout, gs1, gss...)
+end
+
+function Base.map!(f, gsout::Grads, gss::Grads...)
+  all(issetequal(gsout.params, gs.params) for gs in gss) || 
+    throw(ArgumentError("map! expects Grads objects with the same Params."))
+  for p in gsout.params
+    gsout[p] = f((_getformap(gs, p) for gs in gss)...) 
+  end
+  return gsout
+end
+
+function _getformap(gs::Grads, p)
+  g = gs[p]
+  isnothing(g) ? fill!(similar(p), 0) : g 
```

Comment:
Aside: this will likely not be needed one we finish #603 
because the output will be a subtype of `AbstractZero`, rather than `nothing`
Which should be number/array-like enough for all normal use-cases.